src/main/java/de/tum/in/www1/artemis/repository/CompetencyRepository.java

```diff
@@ -60,15 +61,42 @@ public interface CompetencyRepository extends JpaRepository<Competency, Long>, J
             """)
     Optional<Competency> findByIdWithLectureUnitsAndCompletions(@Param("competencyId") long competencyId);
 
+    /**
+     * Fetches all information related to the calculation of the mastery for exercises in a competency.
+     * The complex grouping by is necessary for postgres
+     *
+     * @param competencyId the id of the competency for which to fetch the exercise information
+     * @return the exercise information for the calculation of the mastery in the competency
+     */
+    @Query("""
+            SELECT new de.tum.in.www1.artemis.web.rest.dto.metrics.CompetencyExerciseMasteryCalculationDTO(
+                ex.maxPoints,
+                ex.difficulty,
+                CASE WHEN TYPE(ex) = ProgrammingExercise THEN TRUE ELSE FALSE END,
+                COALESCE(sS.lastScore, tS.lastScore),
+                COALESCE(sS.lastPoints, tS.lastPoints),
+                COALESCE(sS.lastModifiedDate, tS.lastModifiedDate),
+                COUNT(s)
+            )
+            FROM Competency c
+                LEFT JOIN c.exercises ex
+                LEFT JOIN ex.studentParticipations sp
+                LEFT JOIN sp.submissions s
+                LEFT JOIN StudentScore sS ON sS.exercise = ex
+                LEFT JOIN TeamScore tS ON tS.exercise = ex
+            WHERE c.id = :competencyId
+                AND ex IS NOT NULL
+            GROUP BY ex.maxPoints, ex.difficulty, TYPE(ex), sS.lastScore, tS.lastScore, sS.lastPoints, tS.lastPoints, sS.lastModifiedDate, tS.lastModifiedDate
```

Comment:
The complexity increased due to Postgres not liking simpler queries. If someone else finds a way to retrieve the data in an easier way that also works on Postgres, I'll happily use it